src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            ChainedHeader currentHeader = chainedHeader;
+            while (true)
+            {
+                bool headerHasNextHeader = currentHeader.Next.Count > 0;
+                if (headerHasNextHeader)
+                {
+                    this.logger.LogTrace("Header '{0}' is part of another branch.", currentHeader);
+                    break;
+                }
+
+                bool headerHasPeerClaim = this.peerIdsByTipHash.ContainsKey(chainedHeader.HashBlock);
+                if (headerHasPeerClaim)
+                {
+                    this.logger.LogTrace("Header '{0}' is claimed by a peer and wont be removed.", currentHeader);
+                    break;
+                }
+
+                this.logger.LogTrace("Header '{0}' was removed from the chain tree.", currentHeader);
+                this.chainedHeadersByHash.Remove(currentHeader.HashBlock);
+                currentHeader.Previous.Next.Remove(currentHeader);
+
+                currentHeader = currentHeader.Previous;
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Remove the peer's tip and all the headers claimed by this peer unless they are also claimed by other peers.
+        /// </summary>
+        /// <param name="chainedHeader">The header where we start walking back the chain from.</param>
+        /// <param name="networkPeerId">The peer id that is removed.</param>
+        private void RemovePeerClaim(int networkPeerId, ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:{1},{2}:'{3}')", nameof(networkPeerId), networkPeerId, nameof(chainedHeader), chainedHeader);
+
+            var listOfPeersClaimingThisHeader = this.peerIdsByTipHash.TryGet(chainedHeader.HashBlock);
+
+            if (listOfPeersClaimingThisHeader == null)
+            {
+                this.logger.LogTrace("(-)[PEER_TIP_NOT_FOUND]");
+                throw new ConsensusException("PEER_TIP_NOT_FOUND");
+            }
+
+            this.logger.LogTrace("Peer id {0 } removed.", networkPeerId);
+            listOfPeersClaimingThisHeader.Remove(networkPeerId);
+
+            if (listOfPeersClaimingThisHeader.Count == 0)
+            {
+                this.logger.LogTrace("Header '{0}' is not the tip of any peer.", chainedHeader);
+                this.peerIdsByTipHash.Remove(chainedHeader.HashBlock);
+                this.RemoveUnclaimedBranch(chainedHeader);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Set a new header as a tip for this peer and remove the old tip.
+        /// </summary>
+        /// <param name="networkPeerId">The peer id that sets a new tip.</param>
+        /// <param name="newTip">The new tip to set.</param>
+        private void SetPeerTip(int networkPeerId, uint256 newTip)
```

Comment:
`SetPeerTip` => `ClaimPeerTip` / `UpdatePeerTip`

Because it doesnt only set's a new tip, but also removes old claim. So `Set` is a bit missleading

Reply:
I would go with update

Reply:
Or add or replace

Reply:
`AddOrReplacePeerTip` sounds good.

Reply:
Fixed